jsonrpcservice/__init__.py

```diff
@@ -137,66 +140,73 @@ def parsed_request(self):
+            if not isinstance(data, dict):
+                raise InvalidRequest('Expected an object as request body. Batch requests are not supported.')
 
-    def _prepare(self):
-        if self._prepared:
-            return
+            self._dict = data
 
-        self._prepared = True
+        return self._dict
 
-        data = self.parsed_request()
+    @property
+    def id(self):
+        return self.dict.get('id')
 
-        if not isinstance(data, dict):
-            raise InvalidRequest('Expected an object as request body. Batch requests are not supported.')
+    @property
+    def version(self):
+        try:
+            return self.dict['jsonrpc']
+        except KeyError:
+            raise InvalidRequest('Missing "jsonrpc" member in request.')
 
-        # Remember `id` as the first thing. `id` can be ommited for notifications.
-        self._id = data.get('id')
+    @property
+    def method(self):
+        try:
+            method = self.dict['method']
+        except KeyError:
+            raise InvalidRequest('Missing "method" member in request.')
 
-        if 'jsonrpc' not in data:
-            raise InvalidRequest('Missing "jsonrpc" member in request.')
+        if not isinstance(method, str):
+            raise InvalidRequest('"method" of request must be a string.')
 
-        if data['jsonrpc'] != "2.0":
-            raise InvalidRequest('Only JSON-RPC version "2.0" is supported.')
+        return method
 
-        if self._id is None:
-            # Flag as notification only if client specified protocol correctly and thus should know he can't expect
-            # response if id is not set.
-            self._notification = True
+    @property
+    def params(self):
+        if 'params' not in self.dict:
+            return None
 
-        if 'method' not in data:
-            raise InvalidRequest('Missing "method" member in request.')
+        if isinstance(self.dict['params'], (list, dict)):
+            return self.dict['params']
+        else:
+            raise InvalidRequest('"params" of request must be either object or array if present.')
 
-        if not isinstance(data['method'], str):
-            raise InvalidRequest('"method" of request must be a string.')
+    @property
+    def args(self):
+        if isinstance(self.params, list):
+            return self.params
+        else:
+            return []
 
-        self._method = data['method']
+    @property
+    def kwargs(self):
+        if isinstance(self.params, dict):
+            return self.params
+        else:
+            return {}
 
-        if 'params' in data:
-            if isinstance(data['params'], list):
-                self._args.extend(data['params'])
-            elif isinstance(data['params'], dict):
-                self._kwargs.update(data['params'])
-            else:
-                raise InvalidRequest('"params" of request must be either object or array if present.')
+    @property
+    def notification(self):
```

Comment:
I would prefer not to remove `is_` prefix from this property.

Reply:
could do.